test/retain-history/mzcompose.py

```diff
@@ -0,0 +1,206 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+import time
+from datetime import datetime
+from textwrap import dedent
+
+from materialize.mzcompose.composition import Composition
+from materialize.mzcompose.services.cockroach import Cockroach
+from materialize.mzcompose.services.materialized import Materialized
+from materialize.mzcompose.services.testdrive import Testdrive
+
+SERVICES = [
+    Cockroach(setup_materialize=True),
+    Materialized(propagate_crashes=True, external_cockroach=True),
+    Testdrive(no_reset=True, default_timeout="5s"),
+]
+
+
+def workflow_default(c: Composition) -> None:
+    """Test the retain history feature."""
+    run_test(c)
+
+
+def run_test(c: Composition) -> None:
+    c.up("materialized")
+    c.up("testdrive", persistent=True)
+
+    mv_on_mv1_retention_in_sec = 1
+    mv_on_mv_on_mv1_retention_in_sec = 60
+
+    mz_time0 = fetch_now_from_mz(c)
+
+    c.testdrive(
+        dedent(
+            f"""
+            $ postgres-execute connection=postgres://mz_system:materialize@${{testdrive.materialize-internal-sql-addr}}
+            ALTER SYSTEM SET enable_logical_compaction_window = true
+
+            > CREATE TABLE retain_history_table (key INT, value INT);
+
+            > INSERT INTO retain_history_table VALUES (1, 100), (2, 200);
+
+            > CREATE MATERIALIZED VIEW retain_history_mv1 WITH (RETAIN HISTORY FOR '10s') AS
+                    SELECT * FROM retain_history_table;
+
+            > CREATE MATERIALIZED VIEW retain_history_mv_on_mv1 WITH (RETAIN HISTORY FOR '{mv_on_mv1_retention_in_sec}s') AS
+                    SELECT * FROM retain_history_mv1;
+
+            > CREATE MATERIALIZED VIEW retain_history_mv_on_mv_on_mv1 WITH (RETAIN HISTORY FOR '{mv_on_mv_on_mv1_retention_in_sec}s') AS
+                    SELECT * FROM retain_history_mv_on_mv1;
+
+            > SELECT count(*) FROM retain_history_mv1;
+            2
+            """,
+        )
+    )
+
+    mz_time1 = fetch_now_from_mz(c)
+    test_time1 = datetime.now()
+
+    c.testdrive(
+        dedent(
+            f"""
+            > UPDATE retain_history_table SET value = value + 1;
+            > INSERT INTO retain_history_table VALUES (3, 300);
+
+            > SELECT * FROM retain_history_mv1;
+            1 101
+            2 201
+            3 300
+
+            ! SELECT count(*) FROM retain_history_mv1 AS OF '{mz_time0}'::TIMESTAMP;
+            contains: is not valid for all inputs
+
+            > SELECT count(*) FROM retain_history_mv1 AS OF AT LEAST'{mz_time0}'::TIMESTAMP;
```

Comment:
Could this also contain all the values? Since we only specify the `AT LEAST` I think we can't be sure it's always 0, Mz could, with a later optimization, also show the current state.

Reply:
I guess so. I will try to loosen the comparison condition. Or shall I remove these statements?

Reply:
✅ Relaxed comparison values so that the should succeed also with later timestamps.